pkg/operation/botanist/waiter.go

```diff
@@ -74,35 +76,58 @@ func (b *Botanist) WaitUntilKubeAPIServerIsDeleted(ctx context.Context) error {
 
 // WaitUntilKubeAPIServerReady waits until the kube-apiserver pod(s) indicate readiness in their statuses.
 func (b *Botanist) WaitUntilKubeAPIServerReady(ctx context.Context) error {
-	return retry.UntilTimeout(ctx, 5*time.Second, 300*time.Second, func(ctx context.Context) (done bool, err error) {
+	deployment := &appsv1.Deployment{}
 
-		deploy := &appsv1.Deployment{}
-		if err := b.K8sSeedClient.DirectClient().Get(ctx, kutil.Key(b.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeAPIServer), deploy); err != nil {
+	if err := retry.UntilTimeout(ctx, 5*time.Second, 300*time.Second, func(ctx context.Context) (done bool, err error) {
+		if err := b.K8sSeedClient.DirectClient().Get(ctx, kutil.Key(b.Shoot.SeedNamespace, v1beta1constants.DeploymentNameKubeAPIServer), deployment); err != nil {
 			return retry.SevereError(err)
 		}
-		if deploy.Generation != deploy.Status.ObservedGeneration {
+		if deployment.Generation != deployment.Status.ObservedGeneration {
 			return retry.MinorError(fmt.Errorf("kube-apiserver not observed at latest generation (%d/%d)",
-				deploy.Status.ObservedGeneration, deploy.Generation))
+				deployment.Status.ObservedGeneration, deployment.Generation))
 		}
 
 		replicas := int32(0)
-		if deploy.Spec.Replicas != nil {
-			replicas = *deploy.Spec.Replicas
+		if deployment.Spec.Replicas != nil {
+			replicas = *deployment.Spec.Replicas
 		}
-		if replicas != deploy.Status.UpdatedReplicas {
+		if replicas != deployment.Status.UpdatedReplicas {
 			return retry.MinorError(fmt.Errorf("kube-apiserver does not have enough updated replicas (%d/%d)",
-				deploy.Status.UpdatedReplicas, replicas))
+				deployment.Status.UpdatedReplicas, replicas))
 		}
-		if replicas != deploy.Status.Replicas {
+		if replicas != deployment.Status.Replicas {
 			return retry.MinorError(fmt.Errorf("kube-apiserver deployment has outdated replicas"))
 		}
-		if replicas != deploy.Status.AvailableReplicas {
+		if replicas != deployment.Status.AvailableReplicas {
 			return retry.MinorError(fmt.Errorf("kube-apiserver does not have enough available replicas (%d/%d",
-				deploy.Status.AvailableReplicas, replicas))
+				deployment.Status.AvailableReplicas, replicas))
 		}
 
 		return retry.Ok()
-	})
+	}); err != nil {
+		var retryError *retry.Error
+		if !errors.As(err, &retryError) {
+			return err
+		}
+
+		newestPod, err2 := kutil.NewestPodForDeployment(ctx, b.K8sSeedClient.DirectClient(), deployment)
+		if err2 != nil {
+			return err2
+		}
+		if newestPod == nil {
+			return err
+		}
+
+		logs, err2 := kutil.MostRecentCompleteLogs(ctx, b.K8sSeedClient.Kubernetes().CoreV1().Pods(newestPod.Namespace), newestPod, "kube-apiserver", pointer.Int64Ptr(10))
+		if err2 != nil {
+			return err2
+		}
+
```

Comment:
Can we always return `err` and just log `err2`? Reading the logs should happen at best effort and hiding `err` by `err2` might be confusing for users/operators.

Reply:
Makes sense, I've wrapped `err` with the information in `err2` in f589ade